designdocs/CAPACITY.md

```diff
@@ -0,0 +1,148 @@
+
+  UnStake{ account: T::AccountId, msa_id: MessageSenderId, amount: BalanceOf<T> },
+}
+```
+
+Errors when staking
+
+```rust
+pub enum Error<T> {
+	AlreadyStaked,
+	InsufficientBalance,		
+	InvalidMsa,
+}
+```
+
+**Capacity is replenishable**
+
+Replenishable means that after a certain fixed period, called an epoch, all capacity is replenished. The epoch period is composed of blocks.  Moreover, one unit of Capacity is `10^12` and matches with substrate weight measurement where `10^12` is 1 second of execution time.
+
+Since we want the ability to replenish Capacity the following interface is implement to replenish capacity.
+
+```rust
+
+traits Replenishable {
+   type Balance: Balance;
+   
+	 fn replenish_by_account(msa_id: MessageSenderId, amount: Balance) -> Result<Balance, Eroor> {};
+   fn replenish_all_for_account(msa_id: MessageSenderId) -> Result<Balance, Error>;
+   fn can_replenish(msa_id: MessageSenderId) -> bool;
+}
+```
+
+**How capacity replenishes?**
+
+As mentioned above, epoch periods are composed of blocks. The next epoch period calculated is based on the current epoch fullness, and expands and contracts depending on a threshold. This threshold is configurable and can be called `config::epochThreshold` . Additionally, we can also configure a multiplier that we increase or decrease the next epoch based on the fullness of the current epoch. This is defined as `config::epochUtilizationMultiplier`.
+
+![Untitled](https://user-images.githubusercontent.com/3433442/171747512-651112be-bbdc-4197-96ef-0cd208f702db.png)
+
+The above illustrates two epochs where the second one contracts because network congestion has decreased. As a result of epoch decreasing, capacity is replenish faster. Note that initially epoch periods will be fixed in sized for a period of time.
+
+**How are epochs calculated?**
+
+Epoch are used to manage congestion on the network. As demand increases for network resources the epoch length increases which results in capacity taking longer to replenish. 
+
+Upon the finalization of each block, we can get the total amount of weight used and update the total amount of weight for an epoch
+
+![Untitled](https://user-images.githubusercontent.com/3433442/171747526-566ee44a-194f-47e3-8abb-e2c415ac7fb5.png)
+
+**Transactions fees**
+
+When submitting capacity transactions, before including these transaction in a block they are validated at the transaction pool. The validation implemented doing a [SignedExtension](https://docs.rs/sp-runtime/latest/sp_runtime/traits/trait.SignedExtension.html) that validates that the signed transaction is associated to an MSA that contains a Capacity balance. If the MSA has capacity it then increases the amount of capacity is used. If no Capacity is remaining for the epoch period the transaction remains in the future queue until the next epoch. Once the next epoch starts and Capacity becomes available the transaction is processed.
```

Comment:
It seems like future queue can get flooded if epoc becomes too long and we have a lot of transaction without capacity coming in.

Reply:
@aramikm.  @enddynayn and I were talking about if we needed to add a "target epoch" to capacity transactions. That way we could limit future transactions and drop off ones that are overusing instead of just waiting for the next epoch. Thoughts? Would that help with this concern?

Reply:
Yup. I am working on adding more to this section.

Reply:
Is there ever a condition where the pool would drop messages?